### PR TITLE
Moving upsample from data_movement to pool

### PR DESCRIPTION
**Pull request by shwetankTT**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/9759)

### Problem description
Moving upsample from data_movement to pool folder.

### Checklist 
- [x] Post commit CI passes
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by shwetankTT** [comment]
@ayerofieiev-tt Please review this PR.

File: ttnn/cpp/ttnn/operations/pool/upsample/upsample.hpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#include "ttnn/operations/pool/upsample/device/upsample_op.hpp"
+
+namespace ttnn {
+namespace operations {
+namespace upsample {
+
+struct ExecuteUpSample {
+    static ttnn::Tensor operator()(
+        const ttnn::Tensor& input_tensor,
+        std::variant<int, std::array<int, 2>, std::array<int, 3>, std::array<int, 4>> scale_factor,
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/pool/upsample/upsample.hpp:22]:
```suggestion
        const std::variant<int, std::array<int, 2>, std::array<int, 3>, std::array<int, 4>>& scale_factor,
```

I also think we have tt::tt_metal::Array2D, tt::tt_metal::Array3D, tt::tt_metal::Array4D if you want to use them instead.

**Review by ayerofieiev-tt — COMMENTED** [review]
looks good. thank you1
please don't forget to make a good PR title

File: ttnn/cpp/ttnn/operations/pool/upsample/upsample.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+
+#include "upsample.hpp"
+#include "device/upsample_op.hpp"
+#include "ttnn/run_operation.hpp"
+
+namespace ttnn::operations::upsample {
+
+ttnn::Tensor ExecuteUpSample::operator()(const ttnn::Tensor& input_tensor,
+        std::variant<int, tt::tt_metal::Array2D, tt::tt_metal::Array3D, tt::tt_metal::Array4D> scale_factor,
```
Comment (sjameelTT) [inline ttnn/cpp/ttnn/operations/pool/upsample/upsample.cpp:13]: is this double tab indentation?

File: ttnn/cpp/ttnn/operations/pool/upsample/upsample_pybind.hpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+#include "ttnn/cpp/pybind11/decorators.hpp"
+
+#include "upsample.hpp"
+
+namespace ttnn {
+namespace operations{
+namespace upsample{
+namespace detail {
+
```
Comment (sjameelTT) [inline ttnn/cpp/ttnn/operations/pool/upsample/upsample_pybind.hpp:18]: `ttnn::operations::upsample::detail`

File: ttnn/cpp/ttnn/operations/pool/upsample/upsample_pybind.hpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "ttnn/cpp/pybind11/decorators.hpp"
+
+#include "upsample.hpp"
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/pool/upsample/upsample_pybind.hpp:12]: please split into .hpp and .cpp 
reduce headers here, these includes are heavy.

see ccl/all_gather op as example.

File: ttnn/cpp/ttnn/operations/pool/upsample/upsample_pybind.hpp [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+     * :attr:`scale_factor`: multiplier for spatial size. Has to match input size if it is a tuple.
+     )doc";
+
+    using OperationType = decltype(ttnn::upsample);
+    ttnn::bind_registered_operation(
+        module,
+        ttnn::upsample,
+        doc,
+        ttnn::pybind_arguments_t{
+            py::arg("input_tensor"), py::arg("scale_factor"), py::arg("memory_config") = std::nullopt});
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/pool/upsample/upsample_pybind.hpp:39]: memory_config should be a kw_only arg

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/pool/upsample/upsample_pybind.hpp:39]: I recommend to attempt it in this PR.
If it ends up looking like way too many changes - can be done separately

**Review by ayerofieiev-tt — APPROVED** [review]
overall looks good, lets make the final polish and merge

**Comment by ayerofieiev-tt** [comment]
Please update, add links and update status (check checkboxes)
Screenshot 2024-07-31 at 2 47 14 PM

**Comment by shwetankTT** [comment]
@ayerofieiev-tt Please take a quick look. I have addressed your comment.I will be merging after your comment.

File: ttnn/cpp/ttnn/operations/pool/upsample/upsample.cpp (hunk 2 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/pool/upsample/upsample.cpp:13]: Please fix

**Comment by shwetankTT** [comment]
image
I think i need one more approval. @mywoodstock can you please take a look.

**Review by mywoodstock — APPROVED** [review]

**Comment by shwetankTT** [comment]
@tarafdarTT and @TT-BrianLiu Can you guys take a look. I am not able to merge my changes .

**Review by ntarafdar — APPROVED** [review]

**Pull request merged**